File: Problem_2.java
// Time Complexity : O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
// first take a stack, add if the value is grater then last else pop till the value at top is grater then current. then pop the remaining k values.  then check for the leading 0 and remove it.
// Your code here along with comments explaining your approach
class Solution {
    public String removeKdigits(String num, int k) {
       Stack<Integer> st = new Stack<>();
       if(num.length() == k) return "0";
       for(int i = 0; i < num.length(); i++){
           int n = num.charAt(i) - '0';
           while(!st.isEmpty() && k > 0 && n < st.peek()){
                st.pop();
                k--;
           }
           st.add(n);
       }
       for(int i = 0; i<k; i++) {
        st.pop();
       }
       // build the final string, while removing the leading zeros.
        StringBuilder ret = new StringBuilder();
        boolean leadingZero = true;
        for(int digit: st) {
            if(leadingZero && digit == 0) continue;
            leadingZero = false;
            ret.append(digit);
        }

        /* return the final string  */
        if (ret.length() == 0) return "0";
        return ret.toString();
    }
}

File: problem2.cpp
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :



// Your code here along with comments explaining your approach


class Solution {
public:
    string removeKdigits(string num, int k) {
        int n = num.size();
        stack<char>st;
        st.push(num[0]);
        int i = 0;
        for(i=1;i<n;i++)
        {
            while(!st.empty() &&(k>0)&& ( st.top()>num[i]))
            {
                st.pop();
                k--;
            }
            st.push(num[i]);
        }
        while(k--)
        {
            st.pop();
        }
        
        vector<char>v; 
        while(!st.empty())
        {
            v.push_back(st.top());
            st.pop();
        }
        int m = v.size();
        for(int i = m-1;i>=0;i--)
        {
            if(v[i]=='0') {
                v.pop_back();
            }
            else break;
        }
        m = v.size();
        string s(m,'a');
        for(int i = 0;i<m;i++)
        {
            s[i] = v[m-1-i];
        }
        
        return (s=="")?"0":s;
       
    }
};